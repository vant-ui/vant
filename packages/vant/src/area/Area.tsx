import {
  ref,
  watch,
  computed,
  defineComponent,
  type PropType,
  type ExtractPropTypes,
} from 'vue';

// Utils
import {
  pick,
  extend,
  makeArrayProp,
  makeNumericProp,
  createNamespace,
} from '../utils';
import { pickerSharedProps } from '../picker/Picker';
import { INHERIT_PROPS, INHERIT_SLOTS, formatDataForCascade } from './utils';

// Components
import { Picker } from '../picker';

// Types
import type { AreaList } from './types';

const [name, bem] = createNamespace('area');

const areaProps = extend({}, pickerSharedProps, {
  modelValue: String,
  columnsNum: makeNumericProp(3),
  columnsPlaceholder: makeArrayProp<string>(),
  areaList: {
    type: Object as PropType<AreaList>,
    default: () => ({}),
  },
});

export type AreaProps = ExtractPropTypes<typeof areaProps>;

export default defineComponent({
  name,

  props: areaProps,

  emits: ['change', 'confirm', 'cancel', 'update:modelValue'],

  setup(props, { emit, slots }) {
    const codes = ref<string[]>([]);
    const columns = computed(() => formatDataForCascade(props));
    const onChange = (...args: unknown[]) => emit('change', ...args);
    const onCancel = (...args: unknown[]) => emit('cancel', ...args);
    const onConfirm = (...args: unknown[]) => emit('confirm', ...args);

    watch(
      codes,
      (newCodes) => {
        const lastCode = newCodes.length ? newCodes[newCodes.length - 1] : '';
        if (lastCode && lastCode !== props.modelValue) {
          emit('update:modelValue', lastCode);
        }
      },
      { deep: true }
    );

    watch(
      () => props.modelValue,
      (newCode) => {
        const lastCode = codes.value.length
          ? codes.value[codes.value.length - 1]
          : '';
        if (newCode && newCode !== lastCode) {
          codes.value = [
            `${newCode.slice(0, 2)}0000`,
            `${newCode.slice(0, 4)}00`,
            newCode,
          ].slice(0, +props.columnsNum);
        }
      },
      { immediate: true }
    );

    return () => (
      <Picker
        v-model={codes.value}
        v-slots={pick(slots, INHERIT_SLOTS)}
        class={bem()}
        columns={columns.value}
        onChange={onChange}
        onCancel={onCancel}
        onConfirm={onConfirm}
        {...pick(props, INHERIT_PROPS)}
      />
    );
  },
});
