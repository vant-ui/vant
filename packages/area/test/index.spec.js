import Area from '../';
import demoTest from '../../../test/demo-test';

demoTest(Area);
