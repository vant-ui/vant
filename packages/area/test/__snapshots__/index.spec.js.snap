// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change option 1`] = `
<div class="van-picker van-area">
  <div class="van-hairline--top-bottom van-picker__toolbar">
    <div class="van-picker__cancel">取消</div>
    <!---->
    <div class="van-picker__confirm">确认</div>
  </div>
  <!---->
  <div class="van-picker__columns" style="height: 220px;">
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-hairline--top-bottom van-picker__frame" style="height: 44px;"></div>
  </div>
</div>
`;

exports[`change option 2`] = `
<div class="van-picker van-area">
  <div class="van-hairline--top-bottom van-picker__toolbar">
    <div class="van-picker__cancel">取消</div>
    <!---->
    <div class="van-picker__confirm">确认</div>
  </div>
  <!---->
  <div class="van-picker__columns" style="height: 220px;">
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-hairline--top-bottom van-picker__frame" style="height: 44px;"></div>
  </div>
</div>
`;

exports[`change option 3`] = `
<div class="van-picker van-area">
  <div class="van-hairline--top-bottom van-picker__toolbar">
    <div class="van-picker__cancel">取消</div>
    <!---->
    <div class="van-picker__confirm">确认</div>
  </div>
  <!---->
  <div class="van-picker__columns" style="height: 220px;">
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-hairline--top-bottom van-picker__frame" style="height: 44px;"></div>
  </div>
</div>
`;

exports[`watch code 1`] = `
<div class="van-picker van-area">
  <div class="van-hairline--top-bottom van-picker__toolbar">
    <div class="van-picker__cancel">取消</div>
    <!---->
    <div class="van-picker__confirm">确认</div>
  </div>
  <!---->
  <div class="van-picker__columns" style="height: 220px;">
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-hairline--top-bottom van-picker__frame" style="height: 44px;"></div>
  </div>
</div>
`;

exports[`watch code 2`] = `
<div class="van-picker van-area">
  <div class="van-hairline--top-bottom van-picker__toolbar">
    <div class="van-picker__cancel">取消</div>
    <!---->
    <div class="van-picker__confirm">确认</div>
  </div>
  <!---->
  <div class="van-picker__columns" style="height: 220px;">
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-hairline--top-bottom van-picker__frame" style="height: 44px;"></div>
  </div>
</div>
`;

exports[`watch code 3`] = `
<div class="van-picker van-area">
  <div class="van-hairline--top-bottom van-picker__toolbar">
    <div class="van-picker__cancel">取消</div>
    <!---->
    <div class="van-picker__confirm">确认</div>
  </div>
  <!---->
  <div class="van-picker__columns" style="height: 220px;">
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;">
        <li class="van-ellipsis van-picker-column__item van-picker-column__item--selected">北京市</li>
        <li class="van-ellipsis van-picker-column__item">天津市</li>
      </ul>
    </div>
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-hairline--top-bottom van-picker__frame" style="height: 44px;"></div>
  </div>
</div>
`;

exports[`watch code 4`] = `
<div class="van-picker van-area">
  <div class="van-hairline--top-bottom van-picker__toolbar">
    <div class="van-picker__cancel">取消</div>
    <!---->
    <div class="van-picker__confirm">确认</div>
  </div>
  <!---->
  <div class="van-picker__columns" style="height: 220px;">
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;">
        <li class="van-ellipsis van-picker-column__item van-picker-column__item--selected">北京市</li>
        <li class="van-ellipsis van-picker-column__item">天津市</li>
      </ul>
    </div>
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-picker-column" style="height: 220px;">
      <ul style="line-height: 44px;"></ul>
    </div>
    <div class="van-hairline--top-bottom van-picker__frame" style="height: 44px;"></div>
  </div>
</div>
`;
